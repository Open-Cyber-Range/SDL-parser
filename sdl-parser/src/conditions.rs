use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone)]
pub struct Condition {
    #[serde(rename = "vm-name", alias = "Vm-name", alias = "VM-NAME")]
    pub vm_name: String,
    #[serde(default, alias = "Command", alias = "COMMAND")]
    pub command: Option<String>,
    #[serde(default, alias = "Interval", alias = "INTERVAL")]
    pub interval: Option<u32>,
    #[serde(default, alias = "Library", alias = "LIBRARY")]
    pub library: Option<String>,
}

pub type ConditionMap = HashMap<String, Condition>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_condition_is_parsed() {
        let sdl = r#"
            vm-name: windows-10
            command: executable/path.sh
            interval: 30        
        "#;
        let condition = serde_yaml::from_str::<Condition>(sdl).unwrap();
        insta::assert_debug_snapshot!(condition);
    }

    #[test]
    fn library_condition_is_parsed() {
        let sdl = r#"
            vm-name: green-evelation-machine
            library: digital-library-package      
        "#;
        let condition = serde_yaml::from_str::<Condition>(sdl).unwrap();
        insta::assert_debug_snapshot!(condition);
    }

    #[test]
    fn condition_list_is_parsed() {
        let sdl = r#"
            condition-1:
                vm-name: windows-10
                command: executable/path.sh
                interval: 30
            condition-2:
                vm-name: green-evelation-machine
                library: digital-library-package
            condition-3:
                vm-name: green-evelation-machine
                command: executable/path.sh
                interval: 30
        "#;
        let conditions = serde_yaml::from_str::<ConditionMap>(sdl).unwrap();
        insta::with_settings!({sort_maps => true}, {
                insta::assert_yaml_snapshot!(conditions);
        });
    }
}
