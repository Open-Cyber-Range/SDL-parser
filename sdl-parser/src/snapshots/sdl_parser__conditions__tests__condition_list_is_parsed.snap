---
source: sdl-parser/src/conditions.rs
expression: conditions
---
condition-1:
  vm-name: windows-10
  command: executable/path.sh
  interval: 30
  library: ~
condition-2:
  vm-name: green-evelation-machine
  command: ~
  interval: ~
  library: digital-library-package
condition-3:
  vm-name: green-evelation-machine
  command: executable/path.sh
  interval: 30
  library: ~

